src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenidConnectAuthenticationHandler.cs

```diff
@@ -178,33 +234,61 @@ protected override void ApplyResponseChallenge()
 
             if (Options.ProtocolValidator.RequireNonce)
             {
-                openIdConnectMessage.Nonce = Options.ProtocolValidator.GenerateNonce();
+                message.Nonce = Options.ProtocolValidator.GenerateNonce();
                 if (Options.NonceCache != null)
                 {
-                    Options.NonceCache.AddNonce(openIdConnectMessage.Nonce);
+                    if (!Options.NonceCache.TryAddNonce(message.Nonce))
+                    {
+                        string msg = string.Format(CultureInfo.CurrentCulture, Resources.OIDCH_0033_TryAddNonceFailed, message.Nonce);
+                        if (logLevelError)
```

Comment:
We are going to throw anyway so any perf here is small, but I like to be consistent.